### ISSUE TITLE
构建c++20modules出错 (msys2/gcc12)

### ISSUE THREAD

**Issue by C-pops** [xmake-io/xmake#3185]
### Xmake 版本

2.7.3

### 操作系统版本和架构

windows11 22h2

### 描述问题

![image](https://user-images.githubusercontent.com/102885957/208564381-ab70398d-5b70-40ce-b9a8-01f7cea544da.png)
gcc和xmake是msys2下ucrt64环境的

### 期待的结果

应当构建成功

### 工程配置

![image](https://user-images.githubusercontent.com/102885957/208564473-dba19685-9498-442e-99ab-65c831614d0e.png)
![image](https://user-images.githubusercontent.com/102885957/208564504-278ef885-e5d3-43be-8e6b-288e6d1025cb.png)
![image](https://user-images.githubusercontent.com/102885957/208564519-fc0190f8-2015-456c-a05a-dc59d04cd5eb.png)
![image](https://user-images.githubusercontent.com/102885957/208564540-e7fc94ec-fa12-4daa-b0ac-4459c37e2d04.png)

### 附加信息和错误日志

error: src\mod.mpp: error: failed reading mapper 'build\mapper.txt'
src\mod.mpp:1:8: error: unknown Compiled Module Interface: no such module
    1 | export module mod;
      |        ^~~~~~
src\mod.mpp:2: confused by earlier errors, bailing out
  > in src\mod.mpp

**Comment by C-pops** [comment]
![image](https://user-images.githubusercontent.com/102885957/208565135-0d0851c2-8431-4b21-ae59-08d7e661e4b6.png)

**Comment by waruqi** [comment]
最近 dev 版本做了大量 modules 相关改进，可以先尝试更新到 dev 版本，再试试，不行再提供最新错误信息

xmake update -s dev

**Comment by C-pops** [comment]
![image](https://user-images.githubusercontent.com/102885957/208576908-ef180a7d-29ba-4b28-bcac-a470d7e6f9a0.png)
![image](https://user-images.githubusercontent.com/102885957/208577966-b214e805-b257-419c-88a2-3eae0e954278.png)

**Comment by morganwu** [comment]
搭车问一下，MacOS M1上build官方库modules目录下的例子的时候显示下面的错误，请问什么原因呢。
```
 % cd ~/c/xmake/tests/projects/c++/modules/hello
% xmake
[ 14%]: compiling.module.release hello
error: error: module interface compilation requires '-std=c++20' or '-fmodules-ts'
1 error generated.
  > in src/hello.mpp
```

本地环境如下：
```
% clang -v
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

% xmake --version
xmake v2.7.3+dev.30fbe90, A cross-platform build utility based on Lua

 % xmake f -c -vD
checking for platform ... macosx
checking for architecture ... arm64
checking for Xcode directory ... /Applications/Xcode.app
checking for Codesign Identity of Xcode ... no
checking for SDK version of Xcode for macosx (arm64) ... 13.1
checking for Minimal target version of Xcode for macosx (arm64) ... 13.1
checkinfo: cannot runv(dmd --version), No such file or directory
checking for dmd ... no
checkinfo: cannot runv(ldc2 --version), No such file or directory
checking for ldc2 ... no
checkinfo: cannot runv(gdc --version), No such file or directory
checking for gdc ... no
checkinfo: cannot runv(zig version), No such file or directory
checking for zig ... no
checkinfo: cannot runv(zig version), No such file or directory
checking for zig ... no
checking for clang ... /usr/bin/clang
checking for flags (-fPIC) ... ok
> clang "-fPIC"
checking for flags (clang_modules) ... ok
> clang "-fmodules"
checking for flags (clang_builtin_module_map) ... ok
> clang "-fbuiltin-module-map"
checking for flags (clang_implicit_modules) ... ok
> clang "-fimplicit-modules"
configure
{
    ccache = true
    arch = arm64
    clean = true
    ndk_stdcxx = true
    host = macosx
    kind = static
    plat = macosx
    mode = release
    xcode = /Applications/Xcode.app
    buildir = build
}
```

**Comment by waruqi** [comment]
mac 下的 clang 对 modules 支持基本不可用，得自己安装 llvm ，切到自己安装的 llvm 里面的 clang ，用 brew 装下 llvm，然后切到 llvm 工具链去编译

`xmake f --toolchain=llvm`

**Comment by morganwu** [comment]
试着装了llvm，配置的时候提示
```
error: llvm toolchain not found!

% which llvm-dis
/opt/homebrew/opt/llvm/bin/llvm-dis


% llvm-dis --version
Homebrew LLVM version 15.0.6
  Optimized build.
  Default target: arm64-apple-darwin22.2.0
  Host CPU: apple-m1

```
大致搜了一下代码，感觉应该是把搜索位置写死了，固定搜索/usr/local/Cellar目录，实际上M1环境下homebrew安装目录是/opt/homebrew/Cellar，好像也是改不了，有什么配置可以定义llvm路径不？

**Comment by waruqi** [comment]
--sdk

**Comment by morganwu** [comment]
搞定，多谢！

Sample config file for anyone has the same problem with running C++ modules examples on Mac M1.

``` lua
add_rules("mode.release", "mode.debug")

-- customize llvm location if not found
set_config("sdk", "/opt/homebrew/opt/llvm")

set_languages("c++20")
target("hello")
    set_kind("binary")
    add_files("src/*.cpp", "src/*.mpp")

    -- use llvm instead of Clang
    set_toolchains("llvm")
```

**Comment by rewrking** [comment]
Hello. I'm sure if this ever got resolved, but I came across the MinGW issue of:
```
failed reading mapper ... 
unknown Compiled Module Interface: no such module
``` 
while implementing C++20 modules on my own build system. I'm not sure if this got fixed in xmake or not, but figured I would share the solution just in case:

The module mapper file needs to have LF line endings in order for MinGW GCC to read it. I'm not sure what you'd have to do in lua, but in C++, I had to add the binary openmode, otherwise, the resulting mapper file would enforce CRLF line endings, and produce this issue.

```
std::ofstream(mapper_filename, std::ios_base::binary | std::ios_base::out) << mapper_contents
```
